odoo/addons/base/models/qweb.py

```diff
@@ -397,397 +246,340 @@ def get_template(self, template, options):
-                        ),
-                    args=[], keywords=[],
-                    starargs=None, kwargs=None
-                ),
-                body=[ast.If(
-                    test=ast.BoolOp(
-                        op=ast.Or(),
-                        values=[
-                            ast.Name(id='value', ctx=ast.Load()),
-                            ast.Call(
-                                func=ast.Name(id='isinstance', ctx=ast.Load()),
-                                args=[
-                                    ast.Name(id='value', ctx=ast.Load()),
-                                    ast.Name(id='str', ctx=ast.Load())
-                                ],
-                                keywords=[],
-                                starargs=None, kwargs=None
-                            )
-                        ]
-                    ),
-                    body=[
-                        # append(' ')
-                        self._append(ast.Str(u' ')),
-                        # append(name)
-                        self._append(ast.Name('name', ast.Load())),
-                        # append('="')
-                        self._append(ast.Str(u'="')),
-                        # append(escape(str(...)))
-                        self._append(ast.Call(
-                            func=ast.Name('escape', ast.Load()),
-                            args=[ast.Call(
-                                # stringified because regular Markup processing
-                                # is invalid for attributes
-                                func=ast.Name('str', ast.Load()),
-                                args=[ast.BoolOp( # value or ''
-                                    op=ast.Or(),
-                                    values=[ast.Name('value', ast.Load()), ast.Str('')]
-                                )], keywords=[]
-                            )],
-                            keywords=[]
-                        )),
-                        # append('"')
-                        self._append(ast.Str(u'"')),
-                    ],
-                    orelse=[]
-                )],
-                orelse=[]
-            )
-        ]
+    def _compile_expr(self, expr, raise_on_missing=False):
+        """Transform string coming into a python instruction in textual form by
+        adding the namepaces for the dynamic values.
+        This method tokenize the string and call ``__compile_expr_tokens``
+        method.
+        """
+        readable = io.BytesIO(expr.strip().encode('utf-8'))
+        try:
+            tokens = list(tokenize.tokenize(readable.readline))
+        except tokenize.TokenError:
+            raise ValueError(f"Can not compile expression: {expr}")
+
+        qweb_tokens = []
+        for t in tokens:
+            if t.type not in [tokenize.ENCODING, token.ENDMARKER, token.DEDENT]:
+                qweb_tokens.append({'string': t.string, 'type': t.type, 'start': t.start, 'end': t.end})
```

Comment:
The `list` call is useless since you're pushign everything to a new list anyway, and it seems like a waste of time and space to convert the existing tokens to dicts just to have a custom-ish token: you can just instantiate a `tokenize.TokenInfo` object, and optionally add an entry to `token.tok_name` if you want to create a custom token type and not just smuggle the information in an existing one (reusing an existing one might be safer to not risk conflict with new token types).